### PR TITLE
Canonical event schema

### PR DESCRIPTION
**Pull request by rzats**
https://github.com/snowplow/iglu-central/issues/778

### CODE REVIEW

File: schemas/com.snowplowanalytics.snowplow/event/jsonschema/1-0-0 [hunk 1]
```diff
@@ -0,0 +1,488 @@
...
+		"app_id": {
+			"type": ["string", "null"],
+			"maxLength": 255
+		},
+		"platform": {
+			"type": ["string", "null"],
+			"maxLength": 255
+		},
+		"etl_tstamp": {
+			"type": ["string", "null"]
```
Comment (chuwy) [inline schemas/com.snowplowanalytics.snowplow/event/jsonschema/1-0-0]: Aren't all `tstamp`s in `date-time` format?

File: schemas/com.snowplowanalytics.snowplow/event/jsonschema/1-0-0 [hunk 2]
```diff
@@ -0,0 +1,488 @@
...
+		},
+		"v_etl": {
+			"type": "string",
+			"maxLength": 100
+		},
+		"user_id": {
+			"type": ["string", "null"],
+			"maxLength": 255
+		},
+		"user_ipaddress": {
```
Comment (chuwy) [inline schemas/com.snowplowanalytics.snowplow/event/jsonschema/1-0-0]: I think these "hashable" columns ([link](https://github.com/snowplow/snowplow/commit/4bb64acb0ee5aec8a242338fa34ae3d7e93b6deb#diff-4cff77d747a382e71414088ad196028d)) should be something like `oneOf: [{format: ipv4}, {maxLength: 128}]`

File: schemas/com.snowplowanalytics.snowplow/event/jsonschema/1-0-0 [hunk 3]
```diff
@@ -0,0 +1,488 @@
...
+		},
+		"event_fingerprint": {
+			"type": ["string", "null"],
+			"maxLength": 128
+		},
+		"true_tstamp": {
+			"type": ["string", "null"]
+		}
+	},
+	"patternProperties": {
```
Comment (chuwy) [inline schemas/com.snowplowanalytics.snowplow/event/jsonschema/1-0-0]: I was thinking that we shouldn't have these self-describing properties at all and concentrate only on atomic properties (as we're doing this not for Analytics SDK, but for atomic tables), but right now actually I think maybe its not that bad idea. What do you think, @BenFradet?

File: schemas/com.snowplowanalytics.snowplow/event/jsonschema/1-0-0 [hunk 4]
```diff
@@ -0,0 +1,488 @@
...
+		},
+		"doc_width": {
+			"type": ["integer", "null"]
+		},
+		"doc_height": {
+			"type": ["integer", "null"]
+		},
+		"tr_currency": {
+			"type": ["string", "null"],
+			"maxLength": 3
```
Comment (chuwy) [inline schemas/com.snowplowanalytics.snowplow/event/jsonschema/1-0-0]: I think this should also have `minLength: 3`?

File: schemas/com.snowplowanalytics.snowplow/event/jsonschema/1-0-0 (hunk 3 above)
Comment (chuwy) [reply schemas/com.snowplowanalytics.snowplow/event/jsonschema/1-0-0]: Maybe just make it `additionalProperties: true`?

Comment (chuwy) [reply schemas/com.snowplowanalytics.snowplow/event/jsonschema/1-0-0]: So far we don't want to validate anything using this schema, so I'm relatively relaxed about potential mistake.

Comment (chuwy) [reply schemas/com.snowplowanalytics.snowplow/event/jsonschema/1-0-0]: I think maybe it makes sense to rename it to `atomic` and drop self-describing entities altogether?

File: schemas/com.snowplowanalytics.snowplow/atomic/jsonschema/1-0-0 [hunk 5]
```diff
@@ -0,0 +1,513 @@
...
+			"type": ["string", "null"],
+			"anyOf": [
+				{ "format": "ipv4" },
+				{ "maxLength": 128 }
+			]
+		},
+		"user_fingerprint": {
+			"type": ["string", "null"],
+			"anyOf": [
+				{ "format": "ipv4" },
```
Comment (chuwy) [inline schemas/com.snowplowanalytics.snowplow/atomic/jsonschema/1-0-0]: I mean that `user_ipaddress` has `ipv4` format. Others obivously potentially have different constraints, but still which still can be expressed via `oneOf`

File: schemas/com.snowplowanalytics.snowplow/event/jsonschema/1-0-0 (hunk 3 above)
Comment (BenFradet) [reply schemas/com.snowplowanalytics.snowplow/event/jsonschema/1-0-0]: yup agreed on the last point

File: schemas/com.snowplowanalytics.snowplow/atomic/jsonschema/1-0-0 [hunk 6]
```diff
@@ -0,0 +1,513 @@
...
+		},
+		"user_id": {
+			"type": ["string", "null"],
+			"maxLength": 255
+		},
+		"user_ipaddress": {
+			"type": ["string", "null"],
+			"anyOf": [
+				{ "format": "ipv4" },
+				{ "maxLength": 128 }
```
Comment (BenFradet) [inline schemas/com.snowplowanalytics.snowplow/atomic/jsonschema/1-0-0]: is there an ipv6 format?

File: schemas/com.snowplowanalytics.snowplow/atomic/jsonschema/1-0-0 (hunk 5 above)
Comment (BenFradet) [reply schemas/com.snowplowanalytics.snowplow/atomic/jsonschema/1-0-0]: why does user_fingerprint have an ipv4 format?

File: schemas/com.snowplowanalytics.snowplow/atomic/jsonschema/1-0-0 [hunk 7]
```diff
@@ -0,0 +1,513 @@
...
+			]
+		},
+		"domain_userid": {
+			"type": ["string", "null"],
+			"oneOf": [
+				{ "format": "ipv4" },
```
Comment (BenFradet) [inline schemas/com.snowplowanalytics.snowplow/atomic/jsonschema/1-0-0]: same?

File: schemas/com.snowplowanalytics.snowplow/atomic/jsonschema/1-0-0 [hunk 8]
```diff
@@ -0,0 +1,513 @@
...
+				{ "maxLength": 128 }
+			]
+		},
+		"domain_sessionidx": {
+			"type": ["integer", "null"]
+		},
+		"network_userid": {
+			"type": ["string", "null"],
+			"oneOf": [
+				{ "format": "ipv4" },
```
Comment (BenFradet) [inline schemas/com.snowplowanalytics.snowplow/atomic/jsonschema/1-0-0]: same

File: schemas/com.snowplowanalytics.snowplow/atomic/jsonschema/1-0-0 [hunk 9]
```diff
@@ -0,0 +1,513 @@
...
+			"maxLength": 100
+		},
+		"ip_isp": {
+			"type": ["string", "null"],
+			"maxLength": 100
+		},
+		"ip_organization": {
+			"type": ["string", "null"],
+			"oneOf": [
+				{ "format": "ipv4" },
```
Comment (BenFradet) [inline schemas/com.snowplowanalytics.snowplow/atomic/jsonschema/1-0-0]: same?

File: schemas/com.snowplowanalytics.snowplow/atomic/jsonschema/1-0-0 [hunk 10]
```diff
@@ -0,0 +1,513 @@
...
+				{ "maxLength": 128 }
+			]
+		},
+		"ip_domain": {
+			"type": ["string", "null"],
+			"oneOf": [
+				{ "format": "ipv4" },
```
Comment (BenFradet) [inline schemas/com.snowplowanalytics.snowplow/atomic/jsonschema/1-0-0]: same?

File: schemas/com.snowplowanalytics.snowplow/atomic/jsonschema/1-0-0 [hunk 11]
```diff
@@ -0,0 +1,513 @@
...
+			"maxLength": 500
+		},
+		"dvce_sent_tstamp": {
+			"type": ["string", "null"],
+			"format": "date-time"
+		},
+		"refr_domain_userid": {
+			"type": ["string", "null"],
+			"oneOf": [
+				{ "format": "ipv4" },
```
Comment (BenFradet) [inline schemas/com.snowplowanalytics.snowplow/atomic/jsonschema/1-0-0]: same

File: schemas/com.snowplowanalytics.snowplow/atomic/jsonschema/1-0-0 [hunk 12]
```diff
@@ -0,0 +1,513 @@
...
+			]
+		},
+		"refr_dvce_tstamp": {
+			"type": ["string", "null"],
+			"format": "date-time"
+		},
+		"domain_sessionid": {
+			"type": ["string", "null"],
+			"anyOf": [
+				{ "format": "ipv4" },
```
Comment (BenFradet) [inline schemas/com.snowplowanalytics.snowplow/atomic/jsonschema/1-0-0]: same

File: schemas/com.snowplowanalytics.snowplow/atomic/jsonschema/1-0-0 (hunk 6 above)
Comment (chuwy) [reply schemas/com.snowplowanalytics.snowplow/atomic/jsonschema/1-0-0]: Hey @rzats, Ben is right, there's `ipv6` as well.

**Comment by BenFradet** [comment]
cherry-picked, closing :+1:

**Pull request closed without merging**